scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulerBackupRunner.java

```diff
@@ -0,0 +1,123 @@
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.scheduler.core;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.FileTime;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+import org.apache.log4j.Logger;
+import org.joda.time.DateTime;
+import org.ow2.proactive.core.properties.PASharedProperties;
+import org.ow2.proactive.scheduler.common.util.ZipUtils;
+import org.ow2.proactive.utils.ClasspathUtils;
+
+
+public class SchedulerBackupRunner implements Runnable {
+
+    private static final Logger LOGGER = Logger.getLogger(SchedulerBackupRunner.class);
+
+    private static final String PREFIX = "ProActiveSchedulerBackup_";
+
+    private final List<String> targets;
+
+    private final String destination;
+
+    private final int windowSize;
+
+    private final SchedulingService scheduler;
+
+    public SchedulerBackupRunner(SchedulingService scheduler) {
+        this.scheduler = scheduler;
+        targets = Arrays.asList(PASharedProperties.SERVER_BACKUP_TARGETS.getValueAsString().split(","));
+        destination = PASharedProperties.SERVER_BACKUP_DESTINATION.getValueAsString();
+        windowSize = PASharedProperties.SERVER_BACKUP_WINDOWS.getValueAsInt();
+    }
+
+    @Override
+    public void run() {
+        try {
+            scheduler.freeze();
+
+            while (!noOtherTaskIsRunning()) {
+                Thread.sleep(10000);
+            }
+            removeOldBackups();
+            performBackup();
+        } catch (IOException | InterruptedException e) {
+            LOGGER.error(e);
+        } finally {
+            scheduler.resume();
+        }
+    }
+
+    private boolean noOtherTaskIsRunning() {
+        return scheduler.getJobs().getRunningTasks().isEmpty();
+    }
+
+    private void performBackup() throws IOException {
+        String backupFileName = PREFIX + DateTime.now().toString("yyyy-MM-dd'T'HH:mm") + ".zip";
+        File backupFile = new File(destination, backupFileName);
+        LOGGER.info("Performing backup to " + backupFile);
+        String schedulerHome = ClasspathUtils.findSchedulerHome();
+        String[] foldersToZip = targets.stream()
+                                       .map(target -> (new File(schedulerHome, target)).getAbsolutePath())
+                                       .toArray(String[]::new);
+        ZipUtils.zip(foldersToZip, backupFile, null);
+    }
+
+    private void removeOldBackups() {
+        File fodlerWhereBackups = new File(destination);
+        Stream.of(Objects.requireNonNull(fodlerWhereBackups.listFiles()))
+              .filter(File::isFile)
+              .filter(file -> file.getName().startsWith(PREFIX))
+              .sorted(Comparator.comparingLong(file -> {
+                  try {
```

Comment:
this lamba content could be moved to a private method for code clarity

Reply:
ok